### ISSUE TITLE
[$50] Search issues in RB and WP

### ISSUE THREAD

**Issue by sandhiyakavi** [topcoder-platform/taas-apis#254]
Description:

1. Not able to fetch RB details when searching with ProjectId 
2.  Not able to get WP when searching with ResourceBooking Id. (WP seems to be not getting created)

<img width="562" alt="RB_created" src="https://user-images.githubusercontent.com/55895538/118594234-10f74880-b7e4-11eb-880c-e54b9babc2f6.png">
<img width="535" alt="search RB with projectid" src="https://user-images.githubusercontent.com/55895538/118594243-16549300-b7e4-11eb-983a-8e47a70d322a.png">
<img width="546" alt="search WP with RBid" src="https://user-images.githubusercontent.com/55895538/118594247-181e5680-b7e4-11eb-9548-3c6b0de7b2a2.png">

**Comment by nkumar-topcoder** [comment]
@eisbilir 
Resourcebooking status with `placed` status are not getting displayed in frontend or via Resourcebooking search API.

**For example :**
1.  {{URL}}/resourceBookings?projectId=17333 has 7 RB records with `placed` status but it shows only 4 records now.
2. The reason as we suspect is due to this line but I am not sure of that - can you please check. 
https://github.com/topcoder-platform/taas-apis/blob/c63d47e51d81ba0a8637e4166327c4522bd93fdf/src/services/ResourceBookingService.js#L482
3. i,e May be the ES is search condistion includes to have workperiod  records too ? In our context it will always be empty during initials stage of RB record creation with `placed` status
4. Can you fix this ASAP please. Also, note that check if similar issue is exists anywhere else in the code.

Another issue reported but i am unablee to re-create that, Workperiod not getting created
@maxceem pls add

**Comment by maxceem** [comment]
@eisbilir

- as per our findings if ResourceBooking doesn't have start/end (and as c consequence it doesn't have WPs) it is not returned by the endpoint `/resourceBookings/?projectId={{projectId}}` when we search by `projected` for example.
- another thing to check. It looks like for WP to be returned it needs WP payments. somehow these nested records have become mandatory

> Another issue reported but i am unablee to re-create that, Workperiod not getting created

@narekcat I haven't noticed such an issue.

**Comment by narekcat** [comment]
Hi @maxceem . I can work on this issue, if you assign it to me.

**Comment by maxceem** [comment]
thanks, @narekcat, I prefer that **eisbilir** takes care of this on, as he was the one who implemented this code.
If **eisbilir** could not start within 24 hours, I would assign you.

**Comment by narekcat** [comment]
@maxceem Ok.

**Comment by maxceem** [comment]
@eisbilir, please sign-up with Topcoder X tool<br/><br/>```This is an automated message for maxceem via Topcoder X```

**Comment by eisbilir** [comment]
@maxceem 
PR created #256 
fixed the issue: not able to search RBs which has no nested WP data
not related but I discovered another issue and fixed: WPs are not getting deleted when RB's startDate or endDate updated to null
RB and WP related GET endpoints now work good (for ES and DB) wheter they have nested data or not.

**Comment by sandhiyakavi** [comment]
@maxceem @eisbilir Able to fetch RB details with project ID now. But rate type for RB will be by default 'Weekly' but now displayed as 'monthly' is it by default? (RB added through RCRM)

![image](https://user-images.githubusercontent.com/55895538/118774632-0363c080-b8c1-11eb-9944-a0be7b593223.png)

The Workperiod is getting displayed now but not all WP are getting created. RB with date 05/02/2021 - 05/24/2021 has only one WP with 05/09/2021 - 05/15/2021

![image](https://user-images.githubusercontent.com/55895538/118774984-65bcc100-b8c1-11eb-875a-88c906487ab6.png)


![image](https://user-images.githubusercontent.com/55895538/118774844-38701300-b8c1-11eb-869a-9960ffeb8fc0.png)

**Comment by nkumar-topcoder** [comment]
@eisbilir Can you please take a look at the above issue ?

**Comment by sandhiyakavi** [comment]
@maxceem @eisbilir Please ignore the first comment "rate type for RB displayed as 'Monthly'". It's correct  and working as expected.
cc: @nkumar-topcoder

**Comment by eisbilir** [comment]
@sandhiyakavi @maxceem We can set rate type to `weekly` by default if it's needed.

I couldn't re-produce the second one (not all WBs are getting created)

![Postman 2021-05-19 11 15 32](https://user-images.githubusercontent.com/72204071/118781551-db129c80-b895-11eb-9c0c-561f04ba8c25.png)
![Postman 2021-05-19 11 19 36](https://user-images.githubusercontent.com/72204071/118781567-dea62380-b895-11eb-8e1c-c93f56b92393.png)

**Comment by sandhiyakavi** [comment]
@eisbilir It's not showing in my case . Please check the below sample

https://platform.topcoder-dev.com/taas/myteams/17338/rb/7385b68a-cc5f-47ee-9cca-7114e7a5cc8e

Steps to Reproduce:

1. Placed the resource through RCRM.
2. Edited the Job dates to 16-5-2021-29-5-2021 .
3.  Checked the WP created. has only WP (2021-05-23 -2021-05-29) (Last record shown)
4.  After editing the end date to 4-6-2021, 2 WP's are getting displayed, (2021-05-23 to 2021-05-29) and (2021-05-30 to 2021-06-05)

![image](https://user-images.githubusercontent.com/55895538/118788055-5bed8a80-b8ce-11eb-8919-02ac9f484553.png)
![image](https://user-images.githubusercontent.com/55895538/118788120-6b6cd380-b8ce-11eb-8805-a39728855ac5.png)

**Comment by nkumar-topcoder** [comment]
@eisbilir 
I think the issue is with **ElasticSearch**. where only 1 index are created even though the resource has more than 1 WPs.
 _See the example here_
`projectid = 17333, RBID = efc2a3a7-1e63-4d20-b6b4-8fe69b457e44, Startdate=2021-05-24 and endate=2021-06-17`
**the below is response from ES, Only  **1 Workperiod** is getting indexed.
 ```
{
  "_index": "resource_booking",
  "_type": "_doc",
  "_id": "efc2a3a7-1e63-4d20-b6b4-8fe69b457e44",
  "_version": 6,
  "_score": null,
  "_source": {
    "projectId": 17333,
    "userId": "ab19a53b-0607-4a99-8bdd-f3b0cb552293",
    "jobId": "313ffeea-2622-4057-9ec4-c7c2766631f8",
    "rateType": "weekly",
    "status": "placed",
    "id": "efc2a3a7-1e63-4d20-b6b4-8fe69b457e44",
    "createdBy": "00000000-0000-0000-0000-000000000000",
    "updatedAt": "2021-05-19T09:20:54.712Z",
    "createdAt": "2021-05-19T06:54:06.886Z",
    "startDate": "2021-05-24",
    "endDate": "2021-06-17",
    "memberRate": null,
    "customerRate": null,
    "billingAccountId": null,
    "updatedBy": "98ec2c16-442e-4b61-8ad1-66123ee37d3c",
    "workPeriods": [
      {
        "userHandle": "nkumar1",
        "updatedBy": null,
        "endDate": "2021-06-19",
        "resourceBookingId": "efc2a3a7-1e63-4d20-b6b4-8fe69b457e44",
        "daysWorked": null,
        "createdAt": "2021-05-19T09:20:54.970Z",
        "memberRate": null,
        "createdBy": "00000000-0000-0000-0000-000000000000",
        "customerRate": null,
        "id": "5563f269-03db-4eba-85eb-025beefd3fde",
        "projectId": 17333,
        "startDate": "2021-06-13",
        "paymentStatus": "pending",
        "updatedAt": "2021-05-19T09:20:54.970Z"
      }
    ]
  }
```

The workperiods are getting created at PostgressDB with 4 WPs records, which is correct.
 _The below is from postgres DB._
![image](https://user-images.githubusercontent.com/33625707/118790407-5c2c5c80-b8b3-11eb-8563-8c1f3721b913.png)

**Comment by eisbilir** [comment]
The possible reason is the refresh option of taas-es-processor which is set to "wait_for".
After RB is created, taas-es-processor recieves a create message for each WBs and it needs a refresh between the callings. If it is not refreshed after a calling then the next call will override the previos one. That is why only one WB was created.
This is not an issue for local development, but i think it takes more time to create a WB for ES in cloud service.
For creating WB I'ill set the refresh option to `true`, then it need

**Comment by eisbilir** [comment]
PR created https://github.com/topcoder-platform/taas-es-processor/pull/58

**Comment by maxceem** [comment]
Thanks for the quick PR @eisbilir.

We would try to debug this issue more. My biggest concern is, that per my understanding `wait_for` suppose to force the `await esClient.updateExtra({` here https://github.com/topcoder-platform/taas-es-processor/blob/dev/src/services/WorkPeriodProcessorService.js#L30 to wait until data is updated in ES, so the next call of `const resourceBooking = await esClient.getExtra({` suppose to get updated data.

**Comment by nkumar-topcoder** [comment]
@eisbilir 
Pls check Logs from ES processor. The kafka events are coming for workperiod.create topic.
but, when WP creation happening I see Output arguments : undefined.
in this case only the last WP (out of 4) got successfully created at ES.

[log-events-viewer-result (1).csv](https://github.com/topcoder-platform/taas-apis/files/6508031/log-events-viewer-result.1.csv)
![Uploading image.png…]()

**Comment by maxceem** [comment]
@eisbilir my main concern was, that as per my understanding of the ES Processor logic it suppose to work without `true`.

As per my understanding logic is like this:
1. Start processing WP-create event
2. Get RB from ES
3. Add WP to RB
4. Update RB to ES (`wait_for` should force it to wait until data is refreshed)
5. Start processing next WP-create event
6. Get RB from ES (at this moment I would expect already updated RB, as `Update RB to ES` suppose to wait for refresh before complete - maybe I'm wrong here)
7. Add WP to RB
8. Update RB to ES (`wait_for` should force it to wait until data is refreshed)
9. And so on...

So in general I would expect processing WP-create events sequentially one by one, and each step suppose to wait until the data is refreshed until going to the next step. But it appears that it's not like this.

I've checked other Topcoder processors now and I see that all of them are forcing refresh using `true`:
- Resources API https://github.com/topcoder-platform/resource-processor-es/blob/develop/src/services/ProcessorService.js
- Challenges API
https://github.com/topcoder-platform/challenge-processor-es/blob/develop/src/services/ProcessorService.js

As our ES processor is created using the same code so looks like all the processors have this issue without `true`, so it works not as I expected.

For example, in [Project ES processor](https://github.com/topcoder-platform/project-processor-es I've never come across such an issue. Even though we have a complex document here that holds many kinds of nested objects inside. (We had other issues though, but seems like we solved them). And Project ES Processor doesn't set the `refresh` param at all, which means it uses the default `false` value.

TLDR:
- I'm confused about how ES Processor works then if not as I expected
- we would merge your solution and check if it solves the issue

**Comment by nkumar-topcoder** [comment]
@eisbilir PR58 - topcoder-platform/taas-es-processor#58 This solution didn't help.
the same issue persists

**Comment by maxceem** [comment]
@eisbilir after making it `'true'` @nkumar-topcoder tested it twice but the same issue occurs.
Please, let us know if we could do any testing/debugging to help tackle down this issue.

**Comment by nkumar-topcoder** [comment]
@eisbilir @maxceem 
logs details : for Consumer handler

Refer to the file attached for details.
[log-events-viewer-result (2).csv](https://github.com/topcoder-platform/taas-apis/files/6509224/log-events-viewer-result.2.csv)


```
2021-05-19T19:14:04.595+05:30 | [2021-05-19T13:44:04.595Z] app [34mDEBUG[39m : Consumer handler. Topic: taas.workperiod.create, partition: 5, message set length: 1
-- | --
&nbsp; | 2021-05-19T19:14:04.595+05:30 | [2021-05-19T13:44:04.595Z] app [34mDEBUG[39m : Consumer handler. Topic: taas.workperiod.create, partition: 6, message set length: 1
&nbsp; | 2021-05-19T19:14:04.596+05:30 | [2021-05-19T13:44:04.596Z] app [34mDEBUG[39m : Consumer handler. Topic: taas.workperiod.create, partition: 7, message set length: 1
&nbsp; | 2021-05-19T19:14:04.596+05:30 | [2021-05-19T13:44:04.596Z] app [34mDEBUG[39m : Consumer handler. Topic: taas.workperiod.create, partition: 8, message set length: 1
&nbsp; | 2021-05-19T19:14:04.596+05:30 | [2021-05-19T13:44:04.596Z] app [34mDEBUG[39m : Consumer handler. Topic: taas.resourcebooking.update, partition: 4, message set length: 1
&nbsp; | 2021-05-19T19:14:04.597+05:30 | [2021-05-19T13:44:04.597Z] app [32mINFO[39m : Handle Kafka event message; Topic: taas.workperiod.create; Partition: 5; Offset: 55; Message: {"topic":"taas.workperiod.create","originator":"taas-api","timestamp":"2021-05-19T13:44:04.140Z","mime-type":"application/json","payload":{"resourceBookingId":"d9936c37-d5ba-40e1-adf0-ed791d4a5771","startDate":"2021-05-23","endDate":"2021-05-29","daysWorked":null,"paymentStatus":"pending","projectId":17333,"userHandle":"nkumar1","id":"f719950c-d2c0-470e-81b7-68b16b02144f","createdBy":"00000000-0000-0000-0000-000000000000","updatedAt":"2021-05-19T13:44:04.132Z","createdAt":"2021-05-19T13:44:04.132Z","memberRate":null,"customerRate":null,"updatedBy":null}}.
&nbsp; | 2021-05-19T19:14:04.597+05:30 | [2021-05-19T13:44:04.597Z] app [32mINFO[39m : Handle Kafka event message; Topic: taas.workperiod.create; Partition: 6; Offset: 60; Message: {"topic":"taas.workperiod.create","originator":"taas-api","timestamp":"2021-05-19T13:44:04.137Z","mime-type":"application/json","payload":{"resourceBookingId":"d9936c37-d5ba-40e1-adf0-ed791d4a5771","startDate":"2021-05-30","endDate":"2021-06-05","daysWorked":null,"paymentStatus":"pending","projectId":17333,"userHandle":"nkumar1","id":"fc8354af-231a-498d-99bf-3d501baa2b59","createdBy":"00000000-0000-0000-0000-000000000000","updatedAt":"2021-05-19T13:44:04.129Z","createdAt":"2021-05-19T13:44:04.129Z","memberRate":null,"customerRate":null,"updatedBy":null}}.

```

**Comment by maxceem** [comment]
@eisbilir above I've added one more log here https://github.com/topcoder-platform/taas-es-processor/blob/dev/src/app.js#L156.

And we reproduced the issue again to see the log. What I can see is:
- Consumer handler is called in parallel not sequentially (probably because of different partitions) https://github.com/topcoder-platform/taas-es-processor/blob/dev/src/app.js#L154
- As a result the rest of the code is processed in parallel though it's written with an idea to be processed sequentially

![image](https://user-images.githubusercontent.com/146016/118825101-124b7280-b8c3-11eb-803f-04306c9265d0.png)

Now my biggest confusion is why other processors don't have the same issue like us 🤔

**Comment by maxceem** [comment]
> So, instead of spamming WP create events, sending one bulkcreate for WPs will fix the problem here. I'm on it now.

@eisbilir please, don't implement `bulk` updates for WP for now. I wonder if we could make it work without that.

**Comment by eisbilir** [comment]
@maxceem Okay, different partitions seems the reason of this issue. Even if we implement bulk update, current behavior will cause problems for other endpoints either. It's better to find a way to run events sequentially even if they come from different partitions.

**Comment by maxceem** [comment]
> Okay, different partitions seems the reason of this issue. Even if we implement bulk update, current behavior will cause problems for other endpoints either. It's better to find a way to run events sequentially even if they come from different partitions.

Yeah, this is what I'm afraid of. At the moment we are checking what values we have on DEV/PROD env so we could config them in our local docker here https://github.com/topcoder-platform/taas-apis/blob/dev/local/kafka-client/create-topics.sh#L5 to replicate the issue locally and think how to deal with it.

**Comment by sandhiyakavi** [comment]
@maxceem @nkumar-topcoder @eisbilir The WP's are getting displayed correctly now while both adding/editing the dates by increasing and decreasing the RB's dates.

**Comment by maxceem** [comment]
The fix which was applied is we changed the number of partitions from 10 to 1.

**Comment by maxceem** [comment]
@eisbilir, please sign-up with Topcoder X tool<br/><br/>```This is an automated message for maxceem via Topcoder X```

**Comment by maxceem** [comment]
Challenge https://www.topcoder.com/challenges/d567e429-0e7c-491b-a74c-b34974025c90 has been created for this ticket.<br/><br/>```This is an automated message for maxceem via Topcoder X```

**Comment by maxceem** [comment]
> @maxceem Did you decrease the partitions for all of taas topics or just the ones related to current issue?

@nkumar-topcoder could you, please, help here?

**Comment by nkumar-topcoder** [comment]
@eisbilir - related current issue topic of WP only,
we  decreased kafka partitions from 10 to 1 just for taas.workperiod.create, update and delete topics. 
cc @maxceem

**Comment by sandhiyakavi** [comment]
Verified on Prod Env.

Able to fetch RB details when searching with ProjectId
![image](https://user-images.githubusercontent.com/55895538/122372298-88fb8e80-cf9b-11eb-8e29-345c79503237.png)

Able to get WP when searching with ResourceBooking Id.
![image](https://user-images.githubusercontent.com/55895538/122372615-c6601c00-cf9b-11eb-9e85-e1c58f1173f0.png)

The WP's are getting displayed correctly  while both adding/editing the dates by increasing and decreasing the RB's dates.

**Comment by maxceem** [comment]
Challenge d567e429-0e7c-491b-a74c-b34974025c90 has been cancelled<br/><br/>```This is an automated message for maxceem via Topcoder X```

**Comment by maxceem** [comment]
Paid via https://www.topcoder.com/challenges/ff0f7685-81ed-4d81-9d71-a7747ede91fd

**Comment by maxceem** [comment]
Challenge https://www.topcoder.com/challenges/60e5fd07-7aae-4586-bd3c-7d1d36ce848b has been created for this ticket.<br/><br/>```This is an automated message for maxceem via Topcoder X```
